### PR TITLE
Added functionality to add effort sheet url in project

### PR DESCRIPTION
**Pull request by GauravGusain98**
Targets #909 

### Description

- Added input field from effort sheet URL while creating and updating the project
- Added backend validation for effortsheet URL and display them on the frontend
- Added a button to open the sheet on the new tab

image

### Checklist:

- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: Modules/Project/Http/Controllers/ProjectController.php [hunk 1]
```diff
@@ -77,7 +77,7 @@ public function edit(Project $project)
      * Update the specified resource in storage.
      * @param Request $request
      */
-    public function update(Request $request, Project $project)
+    public function update(ProjectRequest $request, Project $project)
```
Comment (GauravGusain98) [inline Modules/Project/Http/Controllers/ProjectController.php:80]: Add validation rules while editing the project

File: Modules/Project/Http/Requests/ProjectRequest.php [hunk 2]
```diff
@@ -28,6 +28,7 @@ public function rules()
             'client_id' => 'required|integer',
             'status' => 'sometimes|string',
             'project_manager' => 'nullable|string',
+            'effort_sheet_url' => 'nullable|active_url|max:255',
```
Comment (GauravGusain98) [inline Modules/Project/Http/Requests/ProjectRequest.php:31]: added rule for effortseet

File: Modules/Project/Http/Requests/ProjectRequest.php [hunk 3]
```diff
@@ -42,6 +43,8 @@ public function messages()
             'client_project_id.required' => 'Project ID is required',
             'client_project_id.integer' => 'Project ID should be a valid number',
             'invoice_email.email' => 'Email for invoice should a valid email address',
+            'effort_sheet_url.max' => 'Url must be less than 255 character',
+            'effort_sheet_url.active_url' => 'Input url is not valid'
```
Comment (GauravGusain98) [inline Modules/Project/Http/Requests/ProjectRequest.php:47]: custom messages for effortsheet validations

File: Modules/Project/Resources/views/edit.blade.php [hunk 4]
```diff
@@ -4,7 +4,10 @@
 <div class="container" id="view_edit_project">
     <br> <h4 clss="c-pointer" v-on:click="counter += 1">{{ $project->name }} ({{ $project->client_project_id }})</h4>
     <br>
-    
+    <div class="text-danger d-none" id="edit-project-errors">
+        <div>Error in the Input:</div>
+        <ul id="edit-project-error-list"></ul>
+    </div>
```
Comment (GauravGusain98) [inline Modules/Project/Resources/views/edit.blade.php:10]: show error messages on UI. 
Note: This error message display is sown only wile editing as wile editing, the axios call is used due to we have to sow the error messages using js

File: Modules/Project/Resources/views/edit.blade.php [hunk 5]
```diff
@@ -51,8 +54,19 @@
 
         updateProjectForm: async function(formId) {
             let formData = new FormData(document.getElementById(formId));
-            let response = await axios.post('{{ route('project.update', $project) }}', formData);
-            alert('Project information updated successfully');
+            await axios.post('{{ route('project.update', $project) }}', formData)
```
Comment (GauravGusain98) [inline Modules/Project/Resources/views/edit.blade.php:57]: Making post request to submit form

File: Modules/Project/Resources/views/edit.blade.php [hunk 6]
```diff
@@ -51,8 +54,19 @@
...
+                alert('Project information updated successfully');
+            })
+            .catch((error) => {
+                let errors = error.response.data.errors;
+                $('#edit-project-error-list').empty()
+                for (error in errors) {
+                    $('#edit-project-error-list').append("<li class='text-danger ml-2'>" + errors[error] + "</li>");
+                }
+                $('#edit-project-errors').removeClass('d-none')
+            })
```
Comment (GauravGusain98) [inline Modules/Project/Resources/views/edit.blade.php:69]: show and hide error messages

File: Modules/Project/Resources/views/subviews/create-project-details.blade.php [hunk 7]
```diff
@@ -17,11 +17,17 @@
...
             </select>
         </div>
+        <div class="form-group col-md-5">
+            <label for="effort_sheet_url">{{ __('Effort-Sheet Url') }}</label>
+            <input type="url" class="form-control" 
+            name="effort_sheet_url" 
+            id="effort_sheet_url" 
+            placeholder="Enter Effortsheet url" 
+            value="{{ old('effort_sheet_url') }}">
+        </div>
```
Comment (GauravGusain98) [inline Modules/Project/Resources/views/subviews/create-project-details.blade.php:27]: added input field for effortsheet url

File: Modules/Project/Services/ProjectService.php [hunk 8]
```diff
@@ -34,6 +34,7 @@ public function store($data)
             'status' => 'active',
             'start_date' => date('Y-m-d'),
             'end_date' => date('Y-m-d'),
+            'effort_sheet_url' => $data['effort_sheet_url'] ?? null
```
Comment (GauravGusain98) [inline Modules/Project/Services/ProjectService.php:37]: store effort sheet url while creating the project

File: Modules/Project/Services/ProjectService.php [hunk 9]
```diff
@@ -83,6 +84,7 @@ private function updateProjectDetails($data, $project)
             'status' => $data['status'],
             'start_date' => date('Y-m-d'),
             'end_date' => date('Y-m-d'),
+            'effort_sheet_url' => $data['effort_sheet_url'] ?? null
```
Comment (GauravGusain98) [inline Modules/Project/Services/ProjectService.php:87]: store effort sheet url while editing the project

File: database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('projects', function (Blueprint $table) {
+            $table->string('effort_sheet_url')->nullable()->after('status');
```
Comment (GauravGusain98) [inline database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php:17]: add `effort_sheet_url` column in `projects` table

File: Modules/Project/Http/Requests/ProjectRequest.php [hunk 11]
```diff
@@ -42,6 +43,8 @@ public function messages()
             'client_project_id.required' => 'Project ID is required',
             'client_project_id.integer' => 'Project ID should be a valid number',
             'invoice_email.email' => 'Email for invoice should a valid email address',
+            'effort_sheet_url.max' => 'Url must be less than 191 character',
```
Comment (kuldeep3) [inline Modules/Project/Http/Requests/ProjectRequest.php:46]:
```suggestion
            'effort_sheet_url.max' => 'Url must be less than 191 characters',
```

File: database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        });
+    }
+
+    /**
+     * Reverse the migrations.
```
Comment (kuldeep3) [inline database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php:22]:
```suggestion
     * Reverse the migration.
```

File: database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class AddEffortSheetUrlColumnInProjectsTable extends Migration
+{
+    /**
+     * Run the migrations.
```
Comment (kuldeep3) [inline database/migrations/2021_08_28_043719_add_effort_sheet_url_column_in_projects_table.php:10]:
```suggestion
     * Run the migration.
```

**Review by kuldeep3 — CHANGES_REQUESTED** [review]
Please make changes as per our offline conversation

File: Modules/Project/Resources/views/edit.blade.php [hunk 14]
```diff
@@ -51,8 +54,19 @@
...
+            await axios.post('{{ route('project.update', $project) }}', formData)
+            .then((response) => {
+                $('#edit-project-errors').addClass('d-none')
+                alert('Project information updated successfully');
+            })
+            .catch((error) => {
+                let errors = error.response.data.errors;
+                $('#edit-project-error-list').empty()
+                for (error in errors) {
+                    $('#edit-project-error-list').append("<li class='text-danger ml-2'>" + errors[error] + "</li>");
```
[the remark this answers is not in the document]
Comment (GauravGusain98) [reply Modules/Project/Resources/views/edit.blade.php:73]: This will return the key of the error array.

**Review by kuldeep3 — APPROVED** [review]

**Pull request merged**